Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp

```diff
@@ -0,0 +1,169 @@
+  entry_form_layout->setFormAlignment(Qt::AlignHCenter | Qt::AlignTop);
+  device_vid_textbox = new QLineEdit();
+  QSizePolicy sizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::Fixed);
+  sizePolicy.setHorizontalStretch(1);
+  sizePolicy.setVerticalStretch(0);
+  sizePolicy.setHeightForWidth(device_vid_textbox->sizePolicy().hasHeightForWidth());
+  device_vid_textbox->setSizePolicy(sizePolicy);
+
+  entry_form_layout->setWidget(2, QFormLayout::LabelRole, device_vid_textbox);
+
+  device_pid_textbox = new QLineEdit();
+  sizePolicy.setHeightForWidth(device_pid_textbox->sizePolicy().hasHeightForWidth());
+  device_pid_textbox->setSizePolicy(sizePolicy);
+
+  entry_form_layout->setWidget(2, QFormLayout::FieldRole, device_pid_textbox);
+  main_layout->addLayout(entry_form_layout);
+
+  select_label = new QLabel(tr("or select a device"));
+  select_label->setAlignment(Qt::AlignCenter);
+
+  main_layout->addWidget(select_label);
+
+  usb_inserted_devices_list = new QListWidget();
+  m_refresh_devices_timer = new QTimer(this);
+  connect(usb_inserted_devices_list, &QListWidget::currentItemChanged, this,
+          &USBDeviceAddToWhitelistDialog::OnDeviceSelection);
+  connect(m_refresh_devices_timer, &QTimer::timeout, this,
+          &USBDeviceAddToWhitelistDialog::OnRefreshDevicesTimer);
+  m_refresh_devices_timer->start(1000);
+
+  main_layout->addWidget(usb_inserted_devices_list);
+  main_layout->addWidget(whitelist_buttonbox);
+
+  // i18n: VID means Vendor ID (in the context of a USB device)
+  device_vid_textbox->setPlaceholderText(tr("Device VID (e.g., 057e)"));
+  // i18n: PID means Product ID (in the context of a USB device), not Process ID
+  device_pid_textbox->setPlaceholderText(tr("Device PID (e.g., 0305)"));
+}
+
+void USBDeviceAddToWhitelistDialog::RefreshDeviceList()
+{
+  const auto& current_devices = USBUtils::GetInsertedDevices();
+  if (current_devices == m_shown_devices)
+    return;
+  QSignalBlocker blocker(usb_inserted_devices_list);  // unblocks when out of scope.
+  const auto selection_string = usb_inserted_devices_list->currentItem();
+  usb_inserted_devices_list->clear();
+  for (const auto& device : current_devices)
+  {
+    if (SConfig::GetInstance().IsUSBDeviceWhitelisted(device.first))
+      continue;
+    usb_inserted_devices_list->addItem(QString::fromStdString(device.second));
+  }
+
+  usb_inserted_devices_list->setCurrentItem(selection_string);
+
+  m_shown_devices = current_devices;
+}
+
+void USBDeviceAddToWhitelistDialog::AddUSBDeviceToWhitelist()
+{
+  const std::string vid_string = StripSpaces(device_vid_textbox->text().toStdString());
+  const std::string pid_string = StripSpaces(device_pid_textbox->text().toStdString());
+  if (!IsValidUSBIDString(vid_string))
+  {
+    // i18n: Here, VID means Vendor ID (for a USB device).
+    QErrorMessage* error = new QErrorMessage();
+    error->setWindowTitle(tr("USB Whitelist Error"));
+    error->showMessage(tr("The entered VID is invalid."));
```

Comment:
Using `QMessageBox::warning` probably makes more sense for these. QErrorMessage has a "Show this message again" checkbox that lets the user disable future dialogs (which also won't actually do anything, because it's being instantiated over and over again).